ocs-api/main.go

```diff
@@ -349,55 +355,179 @@ func postVoucherHandler(w http.ResponseWriter, r *http.Request) {
-func postRereadAgentInstallHandler(w http.ResponseWriter, r *http.Request) {
-	outils.Verbose("POST /api/rereadagentinstall ...")
+//============= GET /api/orgs/{org-id}/keys =============
+// Reads/returns all of the already created owner public key names
+func getKeysHandler(orgId string, w http.ResponseWriter, r *http.Request) {
+	outils.Verbose("GET /api/orgs/%s/keys ...", orgId)
 
 	// Determine the org id to use for the device, based on various inputs
-	deviceOrgId, httpErr := getDeviceOrgId(r)
+	deviceOrgId, httpErr := getDeviceOrgId(orgId, r)
 	if httpErr != nil {
 		http.Error(w, httpErr.Error(), httpErr.Code)
 		return
 	}
 
-	valuesDir := OcsDbDir + "/v1/values"
-	if authenticated, httpErr := outils.ExchangeAuthenticate(r, CurrentExchangeInternalUrl, deviceOrgId, valuesDir+"/agent-install.crt"); httpErr != nil {
+	if authenticated, _, httpErr := outils.ExchangeAuthenticate(r, CurrentExchangeInternalUrl, deviceOrgId, OcsDbDir+"/v1/values/agent-install.crt"); httpErr != nil {
 		http.Error(w, httpErr.Error(), httpErr.Code)
 		return
 	} else if !authenticated {
 		http.Error(w, "invalid exchange credentials provided", http.StatusUnauthorized)
 		return
 	}
 
-	// Reread agent-install.sh
-	fileName := valuesDir + "/agent-install.sh"
-	if httpErr := getAgentInstallScript(fileName); httpErr != nil {
+	// Read the v1/creds/publicKeys/<org> directory in the db, and then read each <user> sub-dir to get all of the pubic key file names
+	pubKeyDirName := OcsDbDir + "/v1/creds/publicKeys/" + deviceOrgId
+	if err := os.MkdirAll(pubKeyDirName, 0750); err != nil { // in case the sub-dir doesn't even exist yet
+		http.Error(w, "could not create directory "+pubKeyDirName+": "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	userDirs, err := ioutil.ReadDir(filepath.Clean(pubKeyDirName))
+	if err != nil {
+		http.Error(w, "Error reading "+pubKeyDirName+" directory: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	pubKeyMap := make(map[string][]string) // the structure we will return (and will get converted to json)
+	for _, u := range userDirs {
+		if u.IsDir() {
+			uDir := pubKeyDirName + "/" + u.Name()
+			keyFiles, err := ioutil.ReadDir(filepath.Clean(uDir))
+			if err != nil {
+				http.Error(w, "Error reading "+uDir+" directory: "+err.Error(), http.StatusBadRequest)
+				return
+			}
+			keyFileNames := make([]string, 0) // the array of public key file names for this user that we will add
+			for _, f := range keyFiles {
+				if !f.IsDir() {
+					keyFileNames = append(keyFileNames, f.Name())
+				}
+			}
+
+			if len(keyFileNames) > 0 {
+				pubKeyMap[u.Name()] = keyFileNames // add it to the map we will return
+			}
+		}
+	}
+
+	// Send file names to client
+	httpCode := http.StatusOK
+	if len(pubKeyMap) == 0 {
+		httpCode = http.StatusNotFound
+	}
+	outils.WriteJsonResponse(httpCode, w, pubKeyMap)
+}
+
+//============= GET /api/orgs/{org-id}/keys/{key-name} =============
+// Reads/returns an already imported voucher
```

Comment:
Did you mean to say "Reads/returns an already existing public key"?

Reply:
Yes i did, thx for catching. Fixed